### PR TITLE
Clarify docs for Supervisor.start_link/2

### PR DESCRIPTION
**Pull request by eksperimental**
The first change is because the was it is expressed it seems like
only a list of modules, a list of tuples or a list of maps is accepts,
while a list with any of these values is.

The second change mentions that an empty children list will return {:ok, pid}.

### CODE REVIEW

File: lib/elixir/lib/supervisor.ex [hunk 1]
```diff
@@ -532,9 +532,12 @@ defmodule Supervisor do
   @doc """
   Starts a supervisor with the given children.
 
-  The children is a list of modules, two-element tuples with module and
-  arguments or a map with the child specification. A strategy is required
-  to be provided through the `:strategy` option. See
+  `children` is a list that includes every child expressed in of the following forms:
+    - a module
+    - a two-element tuple in the shape of `{module, arguments}` with `arguments` being a keyword-list
+    - a [child specification](`t:child_spec/0`)
```
Comment (josevalim) [inline lib/elixir/lib/supervisor.ex:538]:
```suggestion
  `children` is a list of the following forms:

    * a module where `module.child_spec([])` will be invoked to retrieve
      its child specification

    * a two-element tuple in the shape of `{module, arg}` where `module.child_spec(arg)`
      will be invoked for retrieve its child specification

    * a [child specification](`t:child_spec/0`)
```

File: lib/elixir/lib/supervisor.ex [hunk 2]
```diff
@@ -543,7 +546,7 @@ defmodule Supervisor do
 
   If the supervisor and its child processes are successfully spawned
```
Comment (josevalim) [inline lib/elixir/lib/supervisor.ex:547]:
```suggestion
  If the supervisor and all child processes are successfully spawned
```

File: lib/elixir/lib/supervisor.ex [hunk 3]
```diff
@@ -543,7 +546,7 @@ defmodule Supervisor do
...
   (if the start function of each child process returns `{:ok, child}`,
-  `{:ok, child, info}`, or `:ignore`) this function returns
+  `{:ok, child, info}`, or `:ignore`) or if `children` is an empty list, this function returns
```
Comment (josevalim) [inline lib/elixir/lib/supervisor.ex:549]:
```suggestion
  `{:ok, child, info}`, or `:ignore`), this function returns
```

File: lib/elixir/lib/supervisor.ex (hunk 1 above)
Comment (eksperimental) [reply lib/elixir/lib/supervisor.ex:538]: The idea is of the PR is to make it clear that a list combining these elements is valid, unless it is not and I'm wrong.

File: lib/elixir/lib/supervisor.ex (hunk 3 above)
Comment (eksperimental) [reply lib/elixir/lib/supervisor.ex:549]: By saying that "all child processes are successfully spawned" does not cover the "no child process is spawned".
It would be interpreted as that would return `:error`

File: lib/elixir/lib/supervisor.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/supervisor.ex:538]: I am not sure I read the proposal above as saying it must be "only one of those". Lists in Elixir are heterogeneous, so that is the default assumption in all APIs, unless otherwise state?

File: lib/elixir/lib/supervisor.ex (hunk 3 above)
Comment (josevalim) [reply lib/elixir/lib/supervisor.ex:549]: All the quantifier typically covers none as the truth predicate. `Enum.all?([], & &1) == true`. This is typically how the documentation is written and I am not sure we should denote empty lists exceptions everywhere.

Comment (josevalim) [reply lib/elixir/lib/supervisor.ex:549]: We could also rewrite to say: none of its child fail to start by returning `{:error, reason}`. But both all and none are clear predicates.

File: lib/elixir/lib/supervisor.ex (hunk 1 above)
Comment (eksperimental) [reply lib/elixir/lib/supervisor.ex:538]: You are right.

File: lib/elixir/lib/supervisor.ex (hunk 3 above)
Comment (eksperimental) [reply lib/elixir/lib/supervisor.ex:554]: > "All", the quantifier, typically covers empty as the truth predicate. `Enum.all?([], & &1) == true`.

This may not be totally understood in layman terms. In any case going with the negated version of it (`{:error, reason}` while correct and clear it will make it more complicated I guess.

**Comment by eksperimental** [comment]
I made a final change. This is ready to be reviewed and merged.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
